src/nginx_manager.py

```diff
@@ -0,0 +1,75 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Manage nginx instance."""
+
+from pathlib import Path
+
+import nginx
+
+from state import NginxConfig
+from utilities import execute_command
+
+NGINX_SITES_ENABLED_PATH = Path("/etc/nginx/sites-enabled")
+NGINX_DEFAULT_SITES_PATH = NGINX_SITES_ENABLED_PATH / "default"
+CHARM_MANAGED_NGINX_CONFIG_PATH = NGINX_SITES_ENABLED_PATH / "charm-managed"
+
+
+# Unit test is not valuable as the class is closely coupled with nginx.
+# This should be tested with integration tests.
+class NginxManager:  # pragma: no cover
+    """Manages the Nginx instance."""
+
+    def init(self) -> None:
+        """Initialize the nginx server."""
+        # The install, systemctl enable, and systemctl start are idempotent.
+        execute_command(["sudo", "apt", "install", "nginx", "-yq"])
+        execute_command(["sudo", "systemctl", "enable", "nginx"])
+        execute_command(["sudo", "systemctl", "start", "nginx"])
+
+        NGINX_DEFAULT_SITES_PATH.unlink(missing_ok=True)
+
+    def load(self) -> None:
+        """Start or reload the nginx server."""
+        if self.ready_check():
+            # This is reload the configuration files without interrupting service.
+            execute_command(["sudo", "nginx", "-s", "reload"])
+        else:
+            execute_command(["sudo", "systemctl", "restart", "nginx"])
+
+    def stop(self) -> None:
+        """Stop the nginx server."""
+        execute_command(["sudo", "nginx", "-s", "quit"])
+
+    def update_config(self, configuration: NginxConfig) -> None:
+        """Update the nginx configuration files.
+
+        Args:
+            configuration: The nginx locations configurations.
+        """
+        nginx_config = nginx.Conf()
+        server_config = nginx.Server()
+        for i, config in enumerate(configuration.values()):
```

Comment:
You're iterating over locations, so probably better to name it "location_config", or location